### PR TITLE
Batch encoding and decoding of schema data.

### PR DESCRIPTION
**Pull request by robertwb**
This implements conversion directly from a dict of numpy arrays to a concatinated-row-encoded form, and vice versa.

This is generally 5-20x more efficient for encoding and 2-10x more efficient for decoding, depending on the row types and contents.

    Benchmarks:
    
    tiny_row, BatchRowCoderFalse, 1000 element(s)    : p. element median time cost: 3.79658e-06 sec, relative std: 5.82%
    tiny_row, BatchRowCoderTrue, 1000 element(s)     : p. element median time cost: 3.9053e-07 sec, relative std: 6256.41%
    large_row, BatchRowCoderFalse, 1000 element(s)   : p. element median time cost: 2.4623e-05 sec, relative std: 2.46%
    large_row, BatchRowCoderTrue, 1000 element(s)    : p. element median time cost: 1.91796e-06 sec, relative std: 8.61%
    nullable_row, BatchRowCoderFalse, 1000 element(s): p. element median time cost: 2.3849e-05 sec, relative std: 2.15%
    nullable_row, BatchRowCoderTrue, 1000 element(s) : p. element median time cost: 2.27439e-06 sec, relative std: 28.73%
    diverse_row, BatchRowCoderFalse, 1000 element(s) : p. element median time cost: 9.17304e-06 sec, relative std: 2.60%
    diverse_row, BatchRowCoderTrue, 1000 element(s)  : p. element median time cost: 2.42209e-06 sec, relative std: 10.16%

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[Go tests](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by robertwb** [comment]
R: @TheNeuralBit 

We could use this for the input/output ports over the FnAPI boundary. For now it'd mostly help with cross-language, but in the future this (or a variant) could be useful for the GBK crossings as well.

**Comment by TheNeuralBit** [comment]
Very cool, thank you! I will take a detailed look in the next few days. First some high-level comments:
- This currently isn't used anywhere, right? Any thoughts on how we would plumb this in?
- I think it would be a good idea to exercise this with the integration tests in standard_coders.yaml. In a few cases we have multiple test values (e.g. here: https://github.com/apache/beam/blob/6f25fba3fea1f407c7a008c84709047b7a55be54/model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml#L357-L364). We could concatenate together the encoded representations and verify a batch decode.

**Comment by robertwb** [comment]
On Tue, Sep 13, 2022 at 12:05 PM Brian Hulette ***@***.***>
wrote:

> Very cool, thank you! I will take a detailed look in the next few days.
> First some high-level comments:
>
>    - This currently isn't used anywhere, right? Any thoughts on how we
>    would plumb this in?
>
> Correct, this isn't used yet. I was thinking we could possibly add methods
like "encode_batch" and "decode_batch" to the BatchConverters that would
take a coder, whose default implementation would be to explode/encode (or
vice versa) but could be overridden. The input/output ports would take
advantage of this.

>
>    - I think it would be a good idea to exercise this with the
>    integration tests in standard_coders.yaml. In a few cases we have multiple
>    test values (e.g. here:
>    https://github.com/apache/beam/blob/6f25fba3fea1f407c7a008c84709047b7a55be54/model/fn-execution/src/main/resources/org/apache/beam/model/fnexecution/v1/standard_coders.yaml#L357-L364).
>    We could concatenate together the encoded representations and verify a
>    batch decode.
>
>
Good idea. Will do.

**Comment by robertwb** [comment]
The errors look real--looking into them.

**Comment by TheNeuralBit** [comment]
Run Python PreCommit

**Review by TheNeuralBit — APPROVED** [review]
This is very cool, thank you.

I wonder if we should prefer arrow arrays as the default column type though? That way we can specialize strings as well. I think this is OK as-is though since it's not used anywhere and it's effectively a proof of concept. We can experiment with arrow/strings later.

**Comment by TheNeuralBit** [comment]
FYI: @apilloud and @chamikaramj

Currently this is just theoretical, but with some plumbing we could realize a benefit from this when a Batched DoFn is adjacent to a cross-language transform that uses RowCoder, or a shuffle boundary.

@robertwb: I wonder if it would also be faster to use this even in the row-oriented case? Have you checked the performance of this + batching/unbatching against the standard RowCoder?

File: sdks/python/apache_beam/coders/coder_impl_row_encoders.pyx [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+  cdef bint encode_to_stream(self, size_t index, OutputStream stream) except -1:
+    stream.write_var_int64(self.data[index])
+
+  cdef bint decode_from_stream(self, size_t index, InputStream stream) except -1:
+    self.data[index] = stream.read_var_int64()
+
+Int64Int64RowColumnEncoder.register(schema_pb2.INT64, np.int64().dtype)
+
+
+cdef class BoolRowColumnEncoder(AtomicTypeRowColumnEncoder):
```
[the remark this answers is not in the document]
Comment (robertwb) [reply sdks/python/apache_beam/coders/coder_impl_row_encoders.pyx:149]: This is about encoding schema_pb2.BOOLEAN.

**Comment by robertwb** [comment]
> I wonder if we should prefer arrow arrays as the default column type though? That way we can specialize strings as well. I think this is OK as-is though since it's not used anywhere and it's effectively a proof of concept. We can experiment with arrow/strings later.

I think numpy is a lower common denominator than arrow, but this framework certainly lends itself to registration of column encoders of various types, including ragged, justified, or packed arrow string types. (Technically this currently works for anything supporting the https://docs.python.org/3/c-api/buffer.html , though an explicit lookup on the dtype would need to be generalize.)

And, yes, the primary intent is when batch DoFns are adjacent to cross-SDK boundaries of some type. 

Currently, I don't think there's room for savings in decoding a batch + exploding over decoding individual rows (or vice versa) but the former could be a win if we have a row representation that's simply an index into a batch (though that might simply push some of the cost to attribute access).

**Comment by robertwb** [comment]
Note also that windowing still needs to be taken into account, but with the value only coder this code can still be used directly.

File: sdks/python/apache_beam/coders/coder_impl_row_encoders.pyx (hunk 1 above)
[the remark this answers is not in the document]
Comment (TheNeuralBit) [reply sdks/python/apache_beam/coders/coder_impl_row_encoders.pyx:149]: Oh I see, I was confused by the np.int8 and np.uint8 registrations below, but I see now that's a sort of shunt.

**Pull request merged**